plugins/delete-by-query/src/main/java/org/elasticsearch/deletebyquery/action/TransportDeleteByQueryAction.java

```diff
@@ -0,0 +1,454 @@
+
+        AsyncDeleteByQueryAction(DeleteByQueryRequest request, ActionListener<DeleteByQueryResponse> listener) {
+            this.request = request;
+            this.listener = listener;
+            this.startTime = System.currentTimeMillis();
+            this.timedOut = new AtomicBoolean(false);
+            this.total = new AtomicLong(0L);
+            this.semaphore = new Semaphore(0);
+            this.bulks = new AtomicInteger(0);
+            this.shardFailures = ShardSearchFailure.EMPTY_ARRAY;
+            this.bulkResults = new CopyOnWriteArrayList<>();
+        }
+
+        public void start() {
+            executeScan();
+        }
+
+        void executeScan() {
+            try {
+                final SearchRequest scanRequest = new SearchRequest(request.indices()).types(request.types()).indicesOptions(request.indicesOptions());
+                scanRequest.searchType(SearchType.SCAN).scroll(request.scroll());
+                if (request.routing() != null) {
+                    scanRequest.routing(request.routing());
+                }
+
+                SearchSourceBuilder source = new SearchSourceBuilder().query(request.source()).fields("_routing", "_parent").fetchSource(false).version(true);
+                if (request.size() != null) {
+                    source.size(request.size());
+                }
+                if (request.timeout() != null) {
+                    source.timeout(request.timeout());
+                }
+                scanRequest.source(source);
+
+                logger.debug("executing scan request");
+                searchAction.execute(scanRequest, new ActionListener<SearchResponse>() {
+                    @Override
+                    public void onResponse(SearchResponse searchResponse) {
+                        long hits = searchResponse.getHits().getTotalHits();
+                        logger.debug("scan request executed: found [{}] document(s) to delete", hits);
+                        addShardFailures(searchResponse.getShardFailures());
+
+                        if (hits == 0) {
+                            listener.onResponse(buildResponse());
+                            return;
+                        }
+                        total.set(hits);
+
+                        logger.debug("start scrolling [{}] document(s)", hits);
+                        executeScroll(searchResponse.getScrollId());
+                    }
+
+                    @Override
+                    public void onFailure(Throwable e) {
+                        listener.onFailure(e);
+                    }
+                });
+            } catch (Throwable t) {
+                logger.error("unable to execute the initial scan request of delete by query", t);
+                listener.onFailure(t);
+            }
+        }
+
+        void executeScroll(final String scrollId) {
+            threadPool.generic().execute(new AbstractRunnable() {
```

Comment:
I don't think we need to have it execute on another thread pool, its perfectly fine just to go and execute it right away, it doesn't block on anything